File: Div3-Group1/ReverseLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head==NULL || head->next==NULL)
            return head;
            
        ListNode* present=head;
        ListNode* Next=head->next;
        ListNode* preve=NULL;

        while(present->next!=NULL)
        {
            present->next=preve;
            preve=present;

            if(Next!=NULL)
                present=Next;
            if(Next->next!=NULL)
                Next=Next->next;
            else
                Next==NULL;
        }

            present->next=preve;
            preve=present;

        
        return preve;
    }
};

File: leet-code-solutions/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return head

        if head.next == None:
            return head

        curr = head
        prev = curr

        curr=curr.next
        prev.next=None

        while curr:
            next = None
            if curr.next:
                next = curr.next

            curr.next = prev

            if next == None:
                return curr

            prev =curr
            curr = next
            # print(prev.val," ",curr.val,next.val)

        return curr
